### ISSUE TITLE
Unable to cast object of type 'System.String' to type 'System.Int32' when using FromSqlRaw

### ISSUE THREAD

**Issue by sgabler-solytic** [dotnet/efcore#21286]
I'm running the following code:

```
var query = _businessModelContext.KeylessSite
    .FromSqlRaw(
        @"SELECT s.*,
                 COALESCE(dataSourcesCount.count, 0) AS DataSourceCount,
                 address.City,
                 address.Country,
                 ownerContact.FirstName AS OwnerFirstName,
                 ownerContact.LastName AS OwnerLastName,
                 owner.Email                         AS OwnerEmailAddress
          FROM model.[Site] AS s
                   LEFT JOIN (SELECT ds.DefaultSiteId, COUNT(*) AS count
                              FROM model.[DataSource] AS ds
                              GROUP BY ds.DefaultSiteId) AS dataSourcesCount ON dataSourcesCount.DefaultSiteId = s.Id
                   LEFT JOIN model.[Address] AS address ON address.ID = s.AddressID
                   LEFT JOIN model.[User] AS owner ON owner.ID = s.OwnerID
                   LEFT JOIN model.[Contact] AS ownerContact ON ownerContact.Id = owner.ContactID"
    );

var debug = query.ToList(); // exception pops up here
```

### simplified version

I even tried a simplified version of the query with the same result:

```
var query = _businessModelContext.KeylessSite
    .FromSqlRaw(
        @"SELECT s.*,
                 CAST(5 AS INTEGER) AS DataSourceCount,
                 address.City,
                 address.Country,
                 ownerContact.FirstName AS OwnerFirstName,
                 ownerContact.LastName AS OwnerLastName,
                 owner.Email                         AS OwnerEmailAddress
          FROM model.[Site] AS s
                   LEFT JOIN model.[Address] AS address ON address.ID = s.AddressID
                   LEFT JOIN model.[User] AS owner ON owner.ID = s.OwnerID
                   LEFT JOIN model.[Contact] AS ownerContact ON ownerContact.Id = owner.ContactID"
    );

var debug = query.ToList(); // exception pops up here
```

The query itself works fine when executing on the DB, but for some weird reason I keep getting an exception `Unable to cast object of type 'System.String' to type 'System.Int32'`

I think the reason is the `DataSourceCount`, which is the only Integer value in my model.

### KeyLessSite
```
namespace <redacted>.Entities.BusinessModel.Keyless
{
    public partial class KeylessSite : IEntityTypeConfiguration<KeylessSite>, ISite
    {
        public string City { get; set; }
        public string Country { get; set; }
        public string OwnerFirstName { get; set; }
        public string OwnerLastName { get; set; }
        public string OwnerEmailAddress { get; set; }

// THIS IS PROBABLY THE ISSUE
        public int DataSourceCount { get; set; };
// THIS IS PROBABLY THE ISSUE

        public long Id { get; set; }
        public string ImageUrl { get; set; }
        public decimal? Azimuth { get; set; }
        public decimal? Inclination { get; set; }
        public bool? IsPublic { get; set; }

        public void Configure(EntityTypeBuilder<KeylessSite> entity)
        {
            entity.HasNoKey();
        }
    }
}
```

If I change the type of DataSourceCount to string, then I get the inverse error message: `Unable to cast object of type 'System.Int32' to type 'System.String'`

### Steps to reproduce
- create the KeylessSite entity (as above)
- define the query as above (i recommend the simplified query i posted)
- execute the "toList()" statement and check if an error appears

### Exception Details

```
"message": "Unable to cast object of type 'System.String' to type 'System.Int32'.",
        "stackTrace": "   at Microsoft.Data.SqlClient.SqlBuffer.get_Int32()
           at Microsoft.Data.SqlClient.SqlDataReader.GetInt32(Int32 i)
           at lambda_method(Closure , QueryContext , DbDataReader , ResultContext , Int32[] , ResultCoordinator )
           at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
           at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
           at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
           at HotChocolate.Types.Relay.QueryableConnectionResolver`1.GetEdgesToReturn(IQueryable`1 allEdges, QueryablePagingDetails pagingDetails, Int32& offset)
           at HotChocolate.Types.Relay.QueryableConnectionResolver`1.GetSelectedEdges()
           at HotChocolate.Types.Relay.QueryableConnectionResolver`1.Create()
           at HotChocolate.Types.Relay.QueryableConnectionResolver`1.<HotChocolate.Types.Relay.IConnectionResolver.ResolveAsync>b__8_0()
           at System.Threading.Tasks.Task`1.InnerInvoke()
           at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)
           at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
           --- End of stack trace from previous location where exception was thrown ---
           at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
           at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
           --- End of stack trace from previous location where exception was thrown ---
           at HotChocolate.Types.Relay.QueryableConnectionMiddleware`1.InvokeAsync(IMiddlewareContext context)
           at HotChocolate.AspNetCore.Authorization.AuthorizeMiddleware.InvokeAsync(IDirectiveContext context)
           at HotChocolate.Execution.ExecutionStrategyBase.ExecuteMiddlewareAsync(ResolverContext resolverContext, IErrorHandler errorHandler)",
```

### Further technical details

EF Core version: 3.1.4
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3.1
Operating system: Mac

**Comment by sgabler-solytic** [comment]
Before anyone investigates: I have a theory what might be wrong. Seems that I am using an Enum somewhere that can't be cast correctly. 

I'll update or close this issue accordingly

**Comment by smitpatel** [comment]
@sgabler-solytic - Can you also share all the columns in `s.*` ?

**Comment by sgabler-solytic** [comment]
@smitpatel I think I will just close the issue, as I now found the issue.

I had a column in `s.*` which had a Enum type. I needed to add the following code to make it work:

```
entity
    .Property(e => e.InstallationType)
    .HasConversion(new EnumToStringConverter<SiteInstallationType>());

```

This information was originally omitted in the original description, because I was too sure that the issue was somewhere else. That will hopefully teach me to not jump to conclusions ;)
